cyberdrop_dl/crawlers/ashemaletube.py

```diff
@@ -0,0 +1,170 @@
+                title = self.create_title(title)
+                scrape_item.setup_as_album(title)
+            await self.iter_videos(scrape_item, soup.select(PLAYLIST_VIDEOS_SELECTOR))
+
+    @error_handling_wrapper
+    async def profile(self, scrape_item: ScrapeItem) -> None:
+        return await self.model(scrape_item)
+
+    @error_handling_wrapper
+    async def model(self, scrape_item: ScrapeItem) -> None:
+        async for soup in self.web_pager(scrape_item):
+            if model_name := soup.select_one(USER_NAME_SELECTOR):
+                title = model_name.get_text().strip()
+                title = self.create_title(title)
+                scrape_item.setup_as_profile(title)
+            await self.iter_videos(scrape_item, soup.select(PROFILE_VIDEOS_SELECTOR))
+
+    async def iter_videos(self, scrape_item: ScrapeItem, videos) -> None:
+        for video in videos:
+            link: URL = create_canonical_video_url(video.get("href"))
+            new_scrape_item = scrape_item.create_child(link)
+            await self.video(new_scrape_item)
+            scrape_item.add_children()
+
+    async def web_pager(self, scrape_item: ScrapeItem) -> AsyncGenerator[BeautifulSoup]:
+        """Generator of website pages."""
+        page_url = scrape_item.url
+        while True:
+            async with self.request_limiter:
+                soup: BeautifulSoup = await self.client.get_soup_cffi(self.domain, page_url)
+            next_page = soup.select_one(NEXT_PAGE_SELECTOR)
+            yield soup
+            page_url_str: str | None = next_page.get("href") if next_page else None  # type: ignore
+            if not page_url_str:
+                break
+            page_url = self.parse_url(page_url_str, PRIMARY_BASE_DOMAIN)
+
+    @error_handling_wrapper
+    async def video(self, scrape_item: ScrapeItem) -> None:
+        async with self.request_limiter:
+            soup: BeautifulSoup = await self.client.get_soup_cffi(self.domain, scrape_item.url)
+        if player := soup.select_one(JS_SELECTOR):
+            info: dict[str, str | dict] = parse_player_info(player.text)
+            if info is None:
+                raise ScrapeError(404, origin=scrape_item)
+            if info["hls"]:
+                raise ScrapeError(422, origin=scrape_item)
+
+            if date_added := soup.select_one(DATETIME_SELECTOR):
+                scrape_item.possible_datetime = parse_datetime(date_added.get_text(strip=True))
+
+            video_id: str = scrape_item.url.parts[2]
+            title: str = soup.select_one("title").text.split("- aShemaletube.com")[0].strip()
+            filename, ext = self.get_filename_and_ext(info["url"], assume_ext=".mp4")
+            include_id = f"[{video_id}]" if INCLUDE_VIDEO_ID_IN_FILENAME else ""
+            custom_filename = f"{title} {include_id}[{info['res']}]{ext}"
+            await self.handle_file(URL(info["url"]), scrape_item, filename, ext, custom_filename=custom_filename)
```

Comment:
Parse the link once and reuse it.

Also the download link has an unique key. It is also bound to the IP which means it only works for a limited amount of time. Use it as a `debrid_link` and save the page URL as the video link.

You can remove the video title to get an actual canonical URL. It should just be `/videos/<video_id>`. Override the original URL with the canonical URL before calling `self.handle_file`


```suggestion
            link = self.parse_url(info["url"])
            canonical_url = PRIMARY_BASE_DOMAIN / "videos" / video_id
            scrape_item.url = canonical_url 
            filename, ext = self.get_filename_and_ext(link.name, assume_ext=".mp4")
            include_id = f"[{video_id}]" if INCLUDE_VIDEO_ID_IN_FILENAME else ""
            custom_filename = f"{title} {include_id}[{info['res']}]{ext}"
            await self.handle_file(canonical_url, scrape_item, filename, ext, custom_filename=custom_filename, debrid_link=link)
```